packages/pigeon/lib/cpp_generator.dart

```diff
@@ -0,0 +1,794 @@
+        indent.writeln(
+            'void WriteValue(const flutter::EncodableValue& value, flutter::ByteStreamWriter* stream) const;');
+      });
+      indent.writeScoped(' protected:', '', () {
+        indent.writeln(
+            'flutter::EncodableValue ReadValueOfType(uint8_t type, flutter::ByteStreamReader* stream) const;');
+      });
+    }
+  }, nestCount: 0);
+}
+
+void _writeCodecSource(Indent indent, Api api, Root root) {
+  final String codecName = _getCodecName(api);
+  indent.writeln('$codecName::$codecName() {}');
+  if (getCodecClasses(api, root).isNotEmpty) {
+    indent.write(
+        'flutter::EncodableValue $codecName::ReadValueOfType(uint8_t type, flutter::ByteStreamReader* stream) const');
+    indent.scoped('{', '}', () {
+      indent.write('switch (type) ');
+      indent.scoped('{', '}', () {
+        for (final EnumeratedClass customClass in getCodecClasses(api, root)) {
+          indent.write('case ${customClass.enumeration}: ');
+          indent.writeScoped('', '', () {
+            indent.writeln(
+                'return flutter::CustomEncodableValue(${customClass.name}(std::get<flutter::EncodableMap>(ReadValue(stream))));');
+          });
+        }
+        indent.write('default:');
+        indent.writeScoped('', '', () {
+          indent.writeln(
+              'return flutter::StandardCodecSerializer::ReadValueOfType(type, stream);');
+        });
+      });
+    });
+    indent.writeln('');
+    indent.write(
+        'void $codecName::WriteValue(const flutter::EncodableValue& value, flutter::ByteStreamWriter* stream) const');
+    indent.writeScoped('{', '}', () {
+      indent.write(
+          'if (const flutter::CustomEncodableValue* custom_value = std::get_if<flutter::CustomEncodableValue>(&value))');
+      indent.scoped('{', '}', () {
+        for (final EnumeratedClass customClass in getCodecClasses(api, root)) {
+          indent.write(
+              'if (custom_value->type() == typeid(${customClass.name}))');
+          indent.scoped('{', '}', () {
+            indent.writeln('stream->WriteByte(${customClass.enumeration});');
+            indent.writeln(
+                'WriteValue(std::any_cast<${customClass.name}>(*custom_value).ToEncodableMap(), stream);');
+            indent.writeln('return;');
+          });
+        }
+      });
+      indent.writeln(
+          'flutter::StandardCodecSerializer::WriteValue(value, stream);');
+    });
+  }
+}
+
+void _writeErrorOr(Indent indent) {
+  indent.format('''
+class FlutterError {
+ public:
+\tFlutterError();
+\tFlutterError(std::string arg_code)
+\t\t: code(arg_code) {};
+\tFlutterError(std::string arg_code, std::string arg_message)
+\t\t: code(arg_code), message(arg_message) {};
+\tFlutterError(std::string arg_code, std::string arg_message, std::string arg_details)
+\t\t: code(arg_code), message(arg_message), details(arg_details) {};
+\tstd::string code;
+\tstd::string message;
+\tstd::string details;
```

Comment:
I could change this to EncodableValue, but I just copied what the Java implementation does, and we only ever send strings, so I didn't see the point. I wanted to be as specific as possible.

Reply:
Since Pigeon is intended to be general-purpose, I'd rather we not restrict what people can provide as errors. There are scenarios where sending objects back in details is useful.